com.unity.render-pipelines.universal/Editor/ShaderPreprocessor.cs

```diff
@@ -212,15 +212,18 @@ public void OnProcessShader(Shader shader, ShaderSnippetData snippetData, IList<
                 return;
 
             int prevVariantCount = compilerDataList.Count;
-
+            
+            List<ShaderCompilerData> final = new List<ShaderCompilerData>();
             for (int i = 0; i < compilerDataList.Count; ++i)
             {
-                if (StripUnused(ShaderBuildPreprocessor.supportedFeatures, shader, snippetData, compilerDataList[i]))
+                if (!StripUnused(ShaderBuildPreprocessor.supportedFeatures, shader, snippetData, compilerDataList[i]))
                 {
-                    compilerDataList.RemoveAt(i);
-                    --i;
+                    final.Add(compilerDataList[i]);
                 }
             }
+            compilerDataList.Clear();
+            foreach(var i in final)
+                compilerDataList.Add(i);
```

Comment:
I'm worried about this solution, the list here might still be huge. I've looked at HDRP solution. 
They do it in place with a single list. Instead of removing, they swap `i` with `last item`. Then they decrement `last item index`. At the end, they remove all elements at once from `last item index` to list size. Would you be interested in looking at this solution and comparing performance?

// here's the HDRP code, ignore the inner loop about asset
https://github.com/Unity-Technologies/Graphics/blob/master/com.unity.render-pipelines.high-definition/Editor/BuildProcessors/HDRPPreprocessShaders.cs#L256-L295

Reply:
Actually the size of the list is the final result, I think it can't be reduced further more? it only adds the variants which are not striped, but sure I'd like to take a look at the implementation of HDRP

Reply:
The implementation of HDRP is indeed better, so I modified the PR to adopt HDRP's way for doing it